anthos-bm-gcp-terraform/modules/init/main.tf

```diff
@@ -0,0 +1,82 @@
+locals {
+  ssh_pub_key_template_file = "${path.module}/../../resources/ssh-keys.tpl"
+  ssh_pub_key_file          = format(var.pub_key_path_template, var.hostname)
+  ssh_private_key_file      = format(var.priv_key_path_template, var.hostname)
+  cluster_yaml_file_name    = trimprefix(basename(var.cluster_yaml_path), ".")
+}
+
+resource "tls_private_key" "ssh_key_pair" {
+  algorithm = "RSA"
+  rsa_bits  = 4096
+}
+
+resource "local_file" "temp_ssh_pub_key_file" {
+  filename        = local.ssh_pub_key_file
+  file_permission = "0600"
+  content = templatefile(
+    local.ssh_pub_key_template_file,
+    { ssh_key = chomp(tls_private_key.ssh_key_pair.public_key_openssh) }
+  )
+}
+
+resource "local_file" "temp_ssh_priv_key_file" {
+  filename        = local.ssh_private_key_file
+  file_permission = "0600"
+  content         = chomp(tls_private_key.ssh_key_pair.private_key_pem)
+}
+
+module "gcloud_add_ssh_key_metadata" {
+  source                   = "terraform-google-modules/gcloud/google"
+  version                  = "2.0.3"
+  platform                 = "linux"
+  service_account_key_file = var.credentials_file
+  module_depends_on = [
+    local_file.temp_ssh_pub_key_file,
+    local_file.temp_ssh_priv_key_file
+  ]
+
+  create_cmd_entrypoint = "gcloud"
+  create_cmd_body       = <<EOT
+    compute instances add-metadata ${var.hostname}  \
+    --project ${var.project_id}                                    \
+    --zone ${var.zone}                                             \
+    --metadata-from-file ssh-keys=${local.ssh_pub_key_file}
+  EOT
+}
+
+resource "null_resource" "exec_init_script" {
+  depends_on = [module.gcloud_add_ssh_key_metadata]
+  connection {
+    type        = "ssh"
+    user        = "root"
+    host        = var.publicIp
+    private_key = chomp(tls_private_key.ssh_key_pair.private_key_pem)
+  }
+
+  provisioner "file" {
+    source      = var.cluster_yaml_path
+    destination = "/root/${local.cluster_yaml_file_name}"
+  }
+
+  provisioner "file" {
+    source      = var.init_script
+    destination = "/root/init.sh"
+  }
+
+  provisioner "remote-exec" {
+    inline = [
+      "chmod 0100 /root/init.sh",
```

Comment:
why is this file under root?

Reply:
Since the script runs as root user and `/root` is the `$HOME` of the root user

Reply:
yeah, running script as root is not a good idea. I'd recommend using sudo and preferably not using root filesystem.

Reply:
I have fixed this by adding non-root user and user that account to run the scripts: https://github.com/GoogleCloudPlatform/anthos-samples/pull/11/commits/b87454de1cafc7daf1271c4d3a4d379b73dca7fc